### PR TITLE
Added support for Xiaomi Smart Tower Fan

### LINKED ISSUE #213: Unsupported device: dmaker.fan.p39
Just tried adding my fan in Home Assistant, getting: "Unsupported device: dmaker.fan.p39".
Are there any plans on implementing it, or some generic way of controlling it, seems that others has got it to work, although not officially supported by your integration, but there are little info on how to.

Thanks in advance.

### LINKED ISSUE #223: Add dmaker.fan.p39 (Tower Fan) support
Hi,

Just got my new Xiaomi Tower Fan (dmaker.fan.p39) but failed to control it via HA due to lack of support.
Is there any way someone could integrate that?

Much appreciated :-)!

### PR DESCRIPTION
**Pull request by paranerd**
Added support for Xiaomi Smart Tower Fan (dmaker.fan.p39)

For reference:
API overview: https://home.miot-spec.com/spec/dmaker.fan.p39
Spec: https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:fan:0000A005:dmaker-p39:1

Fixes #213 
Fixes #223

### CODE REVIEW

File: custom_components/xiaomi_miio_fan/fan.py [hunk 1]
```diff
@@ -2101,3 +2133,283 @@ def set_rotate(self, direction: FanMoveDirection):
...
+        return (
+            SUPPORT_DIRECTION
+            | SUPPORT_OSCILLATE
+            | SUPPORT_PRESET_MODE
+            | SUPPORT_SET_SPEED
+        )
+
+    """
+    TODO:
+    - fan.turn_on works, but doesn't go to the declared percentage. This is
```
Comment (syssi) [inline custom_components/xiaomi_miio_fan/fan.py:2170]: Do you address these issues before merging?

**Comment by syssi** [comment]
Could you use `black` to format your changes? Thanks in advance!

File: custom_components/xiaomi_miio_fan/fan.py (hunk 1 above)
Comment (paranerd) [reply custom_components/xiaomi_miio_fan/fan.py]: Cleaned up TODOs in latest commits.
Regarding child lock: as stated, this is accurately reflected in HA but I can't seem to be able to actually change it. Am I missing something? Is there supposed to be a button?

**Comment by paranerd** [comment]
> 

This: https://github.com/microsoft/vscode-black-formatter ?

File: custom_components/xiaomi_miio_fan/fan.py (hunk 1 above)
Comment (syssi) [reply custom_components/xiaomi_miio_fan/fan.py]: There is a service called `xiaomi_miio_fan.fan_set_child_lock_on` and `xiaomi_miio_fan.fan_set_child_lock_off`. Did you try to call the service to change the setting?

**Comment by syssi** [comment]
I don't know the VSCode extension. You could install the pypi package called `black`:

```
pip3 install black
```

And execute the tool at the custom component folder once:

```
cd custom_components/xiaomi_miio_fan
black *.py
```

```diff
diff --git a/custom_components/xiaomi_miio_fan/fan.py b/custom_components/xiaomi_miio_fan/fan.py
index 559480b..e31b8af 100644
--- a/custom_components/xiaomi_miio_fan/fan.py
+++ b/custom_components/xiaomi_miio_fan/fan.py
@@ -78,7 +78,7 @@ MODEL_FAN_P11 = "dmaker.fan.p11"  # Mijia Pedestal Fan
 MODEL_FAN_P15 = "dmaker.fan.p15"  # Pedestal Fan Fan P15
 MODEL_FAN_P18 = "dmaker.fan.p18"  # Mi Smart Standing Fan 2
 MODEL_FAN_P33 = "dmaker.fan.p33"  # Mi Smart Standing Fan Pro 2
-MODEL_FAN_P39 = "dmaker.fan.p39" # Smart Tower Fan
+MODEL_FAN_P39 = "dmaker.fan.p39"  # Smart Tower Fan
 MODEL_FAN_LESHOW_SS4 = "leshow.fan.ss4"
 MODEL_FAN_1C = "dmaker.fan.1c"  # Pedestal Fan Fan 1C
 
@@ -362,9 +362,7 @@ FEATURE_FLAGS_FAN_P33 = (
 )
 
 FEATURE_FLAGS_FAN_P39 = (
-    FEATURE_SET_CHILD_LOCK
-    | FEATURE_SET_OSCILLATION_ANGLE
-    | FEATURE_SET_NATURAL_MODE
+    FEATURE_SET_CHILD_LOCK | FEATURE_SET_OSCILLATION_ANGLE | FEATURE_SET_NATURAL_MODE
 )
 
 SERVICE_SET_BUZZER_ON = "fan_set_buzzer_on"
@@ -2025,6 +2023,7 @@ class FanStatusP33(DeviceStatus):
     def angle(self) -> int:
         return self.data["angle"]
 
+
 class FanP33(MiotDevice):
     mapping = {
         # https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:fan:0000A005:dmaker-p33:1
@@ -2134,6 +2133,7 @@ class FanP33(MiotDevice):
             value = 2
         return self.set_property("motor_control", value)
 
+
 class XiaomiFanP39(XiaomiFanMiot):
     """Representation of a Xiaomi Fan P39."""
 
@@ -2300,7 +2300,8 @@ class FanStatusP39(DeviceStatus):
     @property
     def angle(self) -> int:
         return self.data["angle"]
-    
+
+
 class FanP39(MiotDevice):
     mapping = {
         # https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:fan:0000A005:dmaker-p39:1
```

File: custom_components/xiaomi_miio_fan/fan.py (hunk 1 above)
Comment (paranerd) [reply custom_components/xiaomi_miio_fan/fan.py]: Yes, that worked.
So the functionality itself is intact. I was just looking for a dedicated button.

**Review by syssi — APPROVED** [review]
LGTM!

**Comment by syssi** [comment]
Thanks for your contribution!

**Pull request merged**